Allow use of concise query string

busterc opened the pull request:
These updates allow using a concise query `string`. Passing a query `object`
is still supported for backwards compatability, however, reference to using
the underlying object has been removed from the README to avoid creating
overly verbose usage documentation.

busterc committed:
Allow use of concise query string

These updates allow using a concise query `string`. Passing a query `object`
is still supported for backwards compatability, however, reference to using
the underlying object has been removed from the README to avoid creating
overly verbose usage documentation.

bmullan91 commented:
@busterc This is great, thanks for the PR!

I noticed one issue - the string is space sensitive. This will likely trip a lot of people up, especially if they're coming from graphql which is space insensitive.

Eg: The following test would fail

```js
const objectql = require('objectql');
const source = {
  a: 'a',
  b: {
    c: 'c'
  }
};
const query = '{b{c}}';
const actual = objectql(source, query);
const expected = {
  b: {
    c: 'c'
  }
};
```

busterc committed:
Remove space sensitivity

Now '{a{b{c}}}' _like_ querystrings parse properly; an existing
test was modified for validating.

bmullan91 commented on test/querystring-implementation.js:
```diff
@@ -93,7 +93,7 @@ test('it should operate on other keys in the query object recursively', t => {
     }
   };
-  const query = '{ b { c { d } } }';
+  const query = '{b{c{d}}}';
```
Thanks for the update.
For brevity could you also include tests for:

1. `{a}`
1. `{ a}`
1. `{a }`
1. { a }
1. `{a b{ c d {   e }} }`

busterc replied:
Added tests: https://github.com/bmullan91/objectql/pull/1/commits/78e5176eb923539d4441cffcb37f8b4deb23ae05

bmullan91 commented on README.md:
````diff
@@ -37,7 +37,7 @@ objectql({ a: 'a', b: 'b' }, query); // returns { a: 'a' }
 ```
 
-__query__ must be an object, and only the keys where the value is `true` will be used. If an object is used as a key's value it will apply that object as the `query` param to that part of the source object recursively.
+__query__ must be a concise string of key names, like a 'simple' graphql query.
````
I like that you've updated all the docs to use the new style - but we should keep some documentation describing the object way.

What about:
```
__query__ Can be either a string or an object, see below for examples.
```

Then in the `// valid query key values` docs, we have examples of each query in both string + object form - the rest of the examples can be updated to the new string form like you've done.

bmullan91 commented on README.md:
```diff
@@ -67,14 +67,14 @@ objectql(source, { a: function noop() {} }); // returns {}
-objectql(source, { a: true }); // returns { a: 'a' }
-objectql(source, { b: true }); // returns { b: { c: 'c', d: 'd' } }
-objectql(source, { a: true, b: { c: true } }); // returns { a: 'a', b: { c: 'c' } }
+objectql(source, '{ a }'); // returns { a: 'a' }
```
Here's where we should show how to do the same query as an object / string.

busterc replied:
I think this covers the bases: https://github.com/bmullan91/objectql/pull/1/commits/1a8f088487af11f91615af2c0f2b2fd325e99d84

bmullan91 commented on README.md:
```diff
@@ -67,14 +67,14 @@ objectql(source, { a: function noop() {} }); // returns {}
 ## examples
 
-The query object follows a similar pattern to a 'simple' graphql query, for each key in the query `objectql` will pick the matching key from the source object when the query value is `true`.
+The query object follows a similar pattern to a 'simple' graphql query, for each key in the query `objectql` will pick the matching key from the source object.
```
Worth adding:

```
... In the examples below we'll be using the string graphql-like queries.
```

bmullan91 commented on index.js:
```diff
@@ -1,18 +1,47 @@
+        .replace(/" }/g, '":true }')
+    );
+  } catch (error) {
+    return null;
```
We should log then re-throw the error - these errors will typically occur at parse time and should be fixed or it will lead to other issues.

```js
catch (error) {
  console.error(`Invalid objectql query: ${query}`);
  throw error;
}
```

busterc replied:
I set it to return null to avoid a definite API breaking change. With the current v1 API any `!isObject(query)` simply returns the source unaltered: https://github.com/bmullan91/objectql/blob/master/index.js#L4-L5

Are you certain you want to throw here? 

If so, how do you want to update README doc for:
`// invalid queries`

bmullan91 replied:
Ah this is a tricky one... With the previous API, the worst you could do is have a coarse query, since the interpreter would catch invalid objects.

With the string queries it's much more likely a developer could have an invalid query - and if we only logged an error/warning it could just get ignored, defeating the purpose.

> If so, how do you want to update README doc for:
> // invalid queries

We'd keep the same behaviour for queries that aren't objects or strings.

For invalid string query:

```js
// remove...
objectql(source, ''); // returns source

// add...
// invalid string queries
objectql(source, '') // throws
objectql(source, '{ hello }}') // throws
```

I'm open to other ideas though!

bmullan91 commented on index.js:
```diff
@@ -1,18 +1,47 @@
+  }
+
+  // eslint-disable-next-line no-param-reassign
+  query = parse(query);
```
Let's store this in a separate const and remove the `eslint-disable-next-line no-param-reassign`. Naming things is tricky, what about:

```js
const parsedQuery = parse(query);
```

bmullan91 reviewed (changes requested):
Only a few small things - apart from that this is good to go 🎉

busterc committed:
Remove param reassign

Resolves https://github.com/bmullan91/objectql/pull/1#discussion_r221669065

busterc committed:
Add test for space insensitivity

Resolves https://github.com/bmullan91/objectql/pull/1#discussion_r221659385

busterc committed:
Remove query `string` test

Moving the empty `string` test to `querystring-implementation.js` tests.

busterc committed:
Early return non-string query

Don't attempt to parse any non-string query.

busterc committed:
Throw parse errors

Catches, logs to console.error then re-throws an unparseable query.

busterc committed:
Add test for unparseable query

Throws:
- empty string
- malformed query

busterc committed:
Resolves https://github.com/bmullan91/objectql/pull/1#discussion_r221665424

Updated explanation of query.

busterc committed:
Resolves https://github.com/bmullan91/objectql/pull/1#discussion_r221666151

Added explanation about using query string in examples.

busterc committed:
Resolves https://github.com/bmullan91/objectql/pull/1#discussion_r221665666

Shows how to query with object or string.

busterc committed:
Resolves https://github.com/bmullan91/objectql/pull/1#discussion_r222097903

Documents invalid query string expectations.

bmullan91 commented:
Thanks again @busterc - v2 is now published https://www.npmjs.com/package/objectql.

What kind of things are you building with it out of curiosity?

busterc commented:
Thanks @bmullan91 ... for now I'm going to use it for some miscellaneous internal tooling. I wanted an _.pick() that could handle collections (top level/nested) and use graphql'ish syntax. I almost just used objectql as a dep, but figured you might appreciate baking the concise query string feature into objectql.

🍻
